### PR TITLE
Added info about binding to multiple locations

### LINKED ISSUE #1163: Document new extra bindings support for charms metadata
Need to determine where this belongs.

Many charms use a simple model where a relationship with another charm
also indicates there is a network connection between units of those
services. We have utilized this fact to add a network model that allows
system operators to control where those TCP connections are made by
binding the service relation endpoints onto a network space.

However, some charms specifically use relation endpoints only as a way
to pass configuration around, and the relations don't map directly to
services that are running in that charm and/or networking configuration.
These charms want to be able to express that they have more networking
configuration that an operator wants to control without having yet-
another interface that will never be related to another service.

Juju solves the aforementioned issues by introducing an optional new
section in the charm metadata,yaml. The new section is called "extra-
bindings". Similarly to the peers/provides/requires sections, extra-
bindings contains a list of names, which can be used with 'juju deploy
--bind' like relation names. Unlike relations, you don't have to define
hooks or anything more complex to allow the users of the charm to bind
those names to Juju spaces.

Example: ceph charm's metadata.yaml would look like:

    name: ceph
    summary: Highly scalable distributed storage
    maintainer: OpenStack Charmers <openstack-charmers@lists.ubuntu.com>
    description: |
     Ceph is a distributed storage and network file system designed to provide
     excellent performance, reliability, and scalability.
    tags:
      - openstack
      - storage
      - file-servers
      - misc
    peers:
      mon:
        interface: ceph
    extra-bindings:
      cluster:
      public:
    provides:
      nrpe-external-master:
        interface: nrpe-external-master
        scope: container
    …
    storage:
      osd-devices:
        type: block
        multiple:
          range: 0-
      osd-journal:
        type: block
        multiple:
          range: 0-1

As a user then you can deploy this charm and tell Juju to provide
distinct addresses for it on one or more spaces:

    juju deploy ~/path/to/charm/ceph --bind 'cluster=admin-api public=public-api internal-api'

Units of the ceph service will be deployed on machines which have access
to the "admin-api", "public-api", and "internal-api" spaces in MAAS, a
different network interface and address for each binding.

Then, e.g. in a hook of the same charm, running 'network-get cluster
--primary-address' will only return the correct address - the one coming
from the "admin-api" space.

### PR DESCRIPTION
**Pull request by matthewhelmke**
Fixes #1163 and #1175 
@juju/docs

This is a complex topic with information (I believe appropriately) spread across multiple pages. I would appreciate a careful reading and review for both accuracy and clarity.

### CODE REVIEW

File: src/en/authors-charm-metadata.md [hunk 1]
```diff
@@ -132,12 +132,26 @@ payloads:
 ```
 
 `extra-bindings` represents an extra bindable endpoint that is not a relation.
+These are useful whenyou want to have Juju provide distinct addresses for an
```
Comment (evilnick) [inline src/en/authors-charm-metadata.md]: missing space

File: src/en/charms-bundles.md [hunk 2]
```diff
@@ -223,6 +223,35 @@ other than "0", which is used to represent the bootstrap node.  Leaving the
...
+## Binding endpoints of applications within a bundle
+
+You can configure more complex networks using [spaces](./network-spaces.html)
+and deploy charms with binding, as described in [Deploying applications](./charms-deploying.html).
+Bindings can also be specified for applications within a bundle. To do so,
+add a section to the bundle's YAML file called `bindings`. For example:
+
+```
+...
+mysql:
```
Comment (evilnick) [inline src/en/charms-bundles.md]: if this is a YAML fragment, I don't think it is formatted properly

File: src/en/charms-deploying.md [hunk 3]
```diff
@@ -281,6 +281,46 @@ are supplied, then the remaining units will be assigned as normal to a new,
...
+
+You can configure more complex networks using [spaces](./network-spaces.html). This
+also permits the use of `--bind` to specify which network space individual
+charm endpoints should use. When you bind, the endpoint(s) of the charm will
+have IP addresses from subnets that are part of the specified space.
+
+When `--bind` is not used, all endpoints will use the host machine's preferred
+private address, which you can see using `unit-get private-address`.
+
+Create your space before deploying a charm that attempts to use it.
```
Comment (evilnick) [inline src/en/charms-deploying.md]: could we link to where that is described?

**Comment by evilnick** [comment]
It is probably worth asking Dimiter to have a look over this as well

**Comment by matthewhelmke** [comment]
@dimitern can you please take a look at this? Thank you!

File: src/en/charms-deploying.md (hunk 3 above)
Comment (matthewhelmke) [inline src/en/charms-deploying.md]: The link to the page describing configuring/creating spaces is in the very first sentence. Did you mean something else, or would you like it linked twice?

File: src/en/authors-charm-metadata.md [hunk 4]
```diff
@@ -132,12 +132,26 @@ payloads:
...
 ```
+Will permit you to deploy the charm using `--bind` to deploy on units that have
+access to the "admin-api", "public-api", and "internal-api" spaces with a'
+different network interface and address for each binding, using this:
+
+```bash
+juju deploy ~/path/to/charm/foo --bind "cluster=admin-api public=public-api internal-api"
+```
+
+And running `juju network-get cluster --primary-address` will return only the
```
Comment (dimitern) [inline src/en/authors-charm-metadata.md]: s/juju network-get/network-get/ - it's not a juju cli command, but a hook tool.

File: src/en/charms-bundles.md (hunk 2 above)
Comment (dimitern) [inline src/en/charms-bundles.md]: +1

**Comment by dimitern** [comment]
@matthewhelmke  Looks good - only a couple of minor issues. I've also reviewed the full text of the 3 changed files and noticed "lxc" used in a few places. Those should instead use "lxd" - e.g. "lxd:0".

**Comment by matthewhelmke** [comment]
@dimitern Thank you. I'll fix those now along with the issues @evilnick pointed out.

File: src/en/charms-deploying.md (hunk 3 above)
Comment (matthewhelmke) [inline src/en/charms-deploying.md]: I'm going to move the sentence up to the opening paragraph. It still fits there, but makes the link between the two topics more obvious.

**Comment by evilnick** [comment]
Ok :+1:

**Pull request merged**